lib/signalfx/signalflow/computation.rb

```diff
@@ -1,67 +1,218 @@
 # Copyright (C) 2017 SignalFx, Inc. All rights reserved.
 
-# Represents a SignalFlow computation.  A computation can have a channel
+
+STARTED_STATE = :started
+ABORTED_STATE = :aborted
+COMPLETED_STATE = :completed
+DATA_STREAMING_STATE = :data_streaming
+
+# Represents a SignalFlow computation/job.  A computation can have a channel
 # associated with it, but not necessarily (it could have been detached while
-# the computation is still running).  New instances should only be created by
-# the client.
+# the computation is still running or never attached in the first place).  New
+# instances should only be created by the client and a Computation MUST have a
+# handle.
 class Computation
   attr_accessor :handle
   attr_accessor :channel
+  attr_accessor :state
+  attr_accessor :metadata
 
   def initialize(handle, attach_func, stop_func)
     @handle = handle
     @channel = nil
     @attach_func = attach_func
     @stop_func = stop_func
+    @metadata = {}
+    # We can't have a handle until the job is started so we must be at least
+    # at this state
+    @state = STARTED_STATE
+
+    @pending_messages = Queue.new
+
+    @batch_size_known = false
+    @expected_batch_size = 0
+    @current_batch_data = nil
+    @current_batch_size = nil
   end
 
   def channel=(channel)
     @channel = channel
-
-    # This is done in a thread because of the circular nature of channels and
-    # computations in order to avoid trying to grab the same lock recursively
-    # in the channel.  The channel is set in response to a message on the
-    # channel and we can't iterate messages while processing a message.
-    Thread.new do
-      channel.each_message_async do |msg, detach|
-        extract_metadata_from_message(msg)
-      end
-    end
   end
 
   def attached?
-    @channel && !@channel.detached
+    !@channel.nil? && !@channel.detached
   end
 
-  # Iterates over the messages asynchronously for this computation.  See
-  # {Channel#each_message_async}.
+  # Get the next message in this computation.
+  #
+  # @param timeout_seconds [Float] If a new message does not come within this
+  #   interval, raises a {ChannelTimeout} exception.  Note that this does not
+  #   mean that this function will return within this interval since there may
+  #   be messages received that are part of a larger batch.  If nil, will block
+  #   indefinitely.
+  def next_message(timeout_seconds=nil)
+    raise "Computation #{@handle} is not attached to a channel" unless @channel
+
+    # This intermediate message queue exists because data can come in batches.
+    # Due to the logic to determine these batches, the arrival of a single
+    # message on the channel might release multiple messages.  This is where
```

Comment:
I don't think that's true. The arrival of a message may release zero (if you're assembling a full data batch for a given timestamp), or exactly one message (metadata, event, or finally assembled data batch). But never more than one.

Reply:
There is an edge case when you have a `message` message that comes after the initial batch of data.  In that case, you know that the existing batch is complete and should be released to the iterator.  You could just skip releasing the `message` message like the [Python client does](https://github.com/signalfx/signalfx-python/blob/a90ab3d92ce99534ded44ebf5b5b39526b0ab0bf/signalfx/signalflow/computation.py#L133) and release the batch only, but that goes against the pattern of yielding all other types of messages.  Also, the Python client just silently drops a lot of `message` messages without even using them to update computation instance state.

Reply:
The idea is that `message` messages are information about the computation that should be modeled in the logical representation of the computation in the client library. In the Python client, we will progressively extend the scope of what information from those messages is extracted and reflected in the computation instance state, but we don't want users of the client library to start building too much machinery on top of those messages (and instead focus on using properly modeled information in the computation instance).

Reply:
Ok, I made those hidden to users.